fold/fold.go

```diff
@@ -0,0 +1,895 @@
+	}
+
+	if end-start < 4 {
+		foldContext.W[start][end] = InvalidStructure
+		return foldContext.W[start][end], nil
+	}
+
+	w1, err := W(start+1, end, foldContext)
+	if err != nil {
+		return DefaultStructure, fmt.Errorf("w: subsequence (%d, %d): %w", start, end, err)
+	}
+	w2, err := W(start, end-1, foldContext)
+	if err != nil {
+		return DefaultStructure, fmt.Errorf("w: subsequence (%d, %d): %w", start, end, err)
+	}
+	w3, err := V(start, end, foldContext)
+	if err != nil {
+		return DefaultStructure, fmt.Errorf("w: subsequence (%d, %d): %w", start, end, err)
+	}
+
+	w4 := InvalidStructure
+	for k := start + 1; k < end-1; k++ {
+		w4_test, err := Multibranch(start, k, end, foldContext, false)
+		if err != nil {
+			return DefaultStructure, fmt.Errorf("w: subsequence (%d, %d): %w", start, end, err)
+		}
+
+		if w4_test.Valid() && w4_test.Energy < w4.Energy {
+			w4 = w4_test
+		}
+	}
+
+	wret := minStruct(w1, w2, w3, w4)
+	foldContext.W[start][end] = wret
+	return wret, nil
+}
+
+// Find, store and return the minimum free energy of the structure between start
+// and end.
+//
+// If start and end don't bp, store and return INF.
+// See: Figure 2B of Zuker, 1981
+// Args:
+//
+//		start: The start index
+//		end: The end index (inclusive)
+//	 foldContext: The FoldingContext for this sequence
+//
+// Returns the minimum energy folding structure possible between start and end on seq
+func V(start, end int, foldContext FoldContext) (NucleicAcidStructure, error) {
+	if !foldContext.V[start][end].Equal(DefaultStructure) {
+		return foldContext.V[start][end], nil
+	}
+
+	// the ends must basepair for V(start,end)
+	if foldContext.Energies.Complement[foldContext.Seq[start]] != foldContext.Seq[end] {
+		foldContext.V[start][end] = InvalidStructure
+		return foldContext.V[start][end], nil
+	}
+	// if the basepair is isolated, and the seq large, penalize at 1,600 kcal/mol
+	// heuristic for speeding this up
+	// from https://www.ncbi.nlm.nih.gov/pubmed/10329189
+	isolated_outer := true
+	if start > 0 && end < len(foldContext.Seq)-1 {
+		isolated_outer = foldContext.Energies.Complement[foldContext.Seq[start-1]] != foldContext.Seq[end+1]
+	}
+	isolated_inner := foldContext.Energies.Complement[foldContext.Seq[start+1]] != foldContext.Seq[end-1]
+
+	if isolated_outer && isolated_inner {
+		foldContext.V[start][end] = NucleicAcidStructure{Energy: 1600}
+		return foldContext.V[start][end], nil
+	}
+
+	p := Pair(foldContext.Seq, start, start+1, end, end-1)
+	hp, err := Hairpin(start, end, foldContext)
+	if err != nil {
+		return DefaultStructure, fmt.Errorf("v: subsequence (%d, %d): %w", start, end, err)
+	}
+	e1 := NucleicAcidStructure{Energy: hp, Description: "HAIRPIN:" + p}
+	if end-start == 4 { // small hairpin; 4bp
+		foldContext.V[start][end] = e1
+		foldContext.W[start][end] = e1
+		return foldContext.V[start][end], nil
+	}
+
+	n := len(foldContext.Seq)
+	e2 := NucleicAcidStructure{Energy: math.Inf(1)}
+	for i1 := start + 1; i1 < end-4; i1++ {
```

Comment:
what are these magic numbers for?

Reply:
I think it's again the limit of pentanucleotide or less cannot create a structure

Reply:
There are two more magic `4` that I'm not sure about